### ISSUE TITLE
memory.cache's speed depends on the size of ignored arguments

### ISSUE THREAD

**Issue by louisabraham** [joblib/joblib#926]
Hi,

I have a function with the following prototype:

```
@lib.memory.cache(ignore=["doc_names_to_id"])
def build_index(index_path: Path, index_type: str, space: str, doc_names_to_id: dict):
    print("begin")
```

The time that I have to wait before seeing "begin" depends on the length of `doc_names_to_id`, which I believe is not a feature.
During this time, according to `htop`, the CPU usage is 100% but the memory consumption does not increase.

First, I thought that memory.cache computes the hash of all arguments, including the ones present in the `ignored` argument. After reading the code, I don't think it should be the case.
Also, I confirmed by logging `_get_argument_hash` that it is not the case.

I finally found that the problem comes from _format_arg:

https://github.com/joblib/joblib/blob/master/joblib/func_inspect.py#L321-L325

Basically, the problem is that my dict is so long that pformat is very slow.

I believe nobody had the problem before because numpy arrays format to short strings.

I'm willing to implement a fix, what do you think is the correct way to address that?

**Comment by louisabraham** [comment]
A possible idea would be to truncate the length of the object to 500. As you truncate the length of the final string to 700 or 1500, I think this would not change the result.

However, recursive objects would still be a problem, so I believe a function that limits the length the flattened object would be the answer.

Another (simpler) possibility would be to add an option for non-logged objects.

**Comment by GaelVaroquaux** [comment]
These suggestions sound reasonable to me. Thanks for investigating this!
